File: server/src/main/java/com/uten/imp/features/visitor/VisitorApplication.java
package com.uten.imp.features.visitor;

import com.uten.imp.common.domain.SoftDeletableEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Table;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.OffsetDateTime;
import java.util.UUID;

/**
 * 访客来访申请。状态机：pending → hostReviewing(可选) → approved/rejected → checkedIn。
 */
@Getter
@Setter
@NoArgsConstructor
@Entity
@Table(name = "visitor_applications")
public class VisitorApplication extends SoftDeletableEntity {

    @Column(name = "visitor_account_id")
    private UUID visitorAccountId;

    @Column(name = "visitor_name", nullable = false)
    private String visitorName;

    @Column(name = "phone_enc")
    private String phoneEnc;

    @Column(name = "id_card_enc")
    private String idCardEnc;

    @Column(name = "id_card_last4")
    private String idCardLast4;

    private String company;

    @Column(name = "visit_purpose", nullable = false)
    private String visitPurpose;

    @Column(name = "has_vehicle", nullable = false)
    private boolean hasVehicle;

    @Column(name = "plate_no")
    private String plateNo;

    @Column(name = "plate_no_enc")
    private String plateNoEnc;

    @Column(name = "host_employee_id")
    private UUID hostEmployeeId;

    @Column(name = "host_department_id")
    private UUID hostDepartmentId;

    @Column(name = "planned_visit_at", nullable = false)
    private OffsetDateTime plannedVisitAt;

    @Column(name = "planned_leave_at")
    private OffsetDateTime plannedLeaveAt;

    @Column(nullable = false)
    private String status = "pending";

    @Column(name = "applied_at", nullable = false)
    private OffsetDateTime appliedAt = OffsetDateTime.now();

    @Column(name = "approved_by")
    private UUID approvedBy;

    @Column(name = "approved_at")
    private OffsetDateTime approvedAt;

    @Column(name = "reject_reason")
    private String rejectReason;

    /** 被访人确认（两级审批可选）：null=未走该环节，true=同意，false=拒绝。 */
    @Column(name = "host_confirmed")
    private Boolean hostConfirmed;

    @Column(name = "check_in_at")
    private OffsetDateTime checkInAt;

    /** HR 批准后签发的入场二维码凭证（HMAC 签名）。 */
    @Column(name = "qr_token", unique = true)
    private String qrToken;

    /** 6位短码：二维码无法扫描时保安手动输入核验。 */
    @Column(name = "passcode")
    private String passcode;
}

File: lib/features/visitor/models/visitor_application.dart
// 访客来访申请模型（对应后端 VisitorApplication / VisitorListItem / VisitorDetail）。

/// 申请状态：pending 申请中 / hostReviewing 转被访人 / approved 已批准 /
/// rejected 已拒绝 / checkedIn 已签到 / cancelled 已取消。
enum VisitorApplicationStatus { pending, hostReviewing, approved, rejected, checkedIn, cancelled }

VisitorApplicationStatus visitorStatusFromCode(String? code) {
  switch (code) {
    case 'pending':
      return VisitorApplicationStatus.pending;
    case 'hostReviewing':
      return VisitorApplicationStatus.hostReviewing;
    case 'approved':
      return VisitorApplicationStatus.approved;
    case 'rejected':
      return VisitorApplicationStatus.rejected;
    case 'checkedIn':
      return VisitorApplicationStatus.checkedIn;
    case 'cancelled':
      return VisitorApplicationStatus.cancelled;
    default:
      return VisitorApplicationStatus.pending;
  }
}

class VisitorApplication {
  const VisitorApplication({
    required this.id,
    required this.visitorName,
    required this.visitPurpose,
    required this.status,
    required this.plannedVisitAt,
    required this.appliedAt,
    this.company,
    this.hostName,
    this.hostDepartment,
    this.plannedLeaveAt,
    this.approvedAt,
    this.hasVehicle = false,
    this.plateNo,
    this.rejectReason,
    this.qrToken,
    this.passcode,
  });

  final String id;
  final String visitorName;
  final String visitPurpose;
  final VisitorApplicationStatus status;
  final DateTime plannedVisitAt;
  final DateTime appliedAt;
  final String? company;
  final String? hostName;
  final String? hostDepartment;
  final DateTime? plannedLeaveAt;
  final DateTime? approvedAt;
  final bool hasVehicle;
  final String? plateNo;
  final String? rejectReason;
  final String? qrToken;
  final String? passcode;

  factory VisitorApplication.fromJson(Map<String, dynamic> j) {
    DateTime parse(Object? v) => v is String && v.isNotEmpty
        ? DateTime.tryParse(v) ?? DateTime.now()
        : DateTime.now();
    return VisitorApplication(
      id: (j['id'] ?? '').toString(),
      visitorName: (j['visitorName'] ?? '').toString(),
      visitPurpose: (j['visitPurpose'] ?? '').toString(),
      status: visitorStatusFromCode(j['status'] as String?),
      plannedVisitAt: parse(j['plannedVisitAt']),
      appliedAt: parse(j['appliedAt']),
      company: j['company'] as String?,
      hostName: j['hostName'] as String?,
      hostDepartment: j['hostDepartment'] as String?,
      plannedLeaveAt: j['plannedLeaveAt'] == null ? null : parse(j['plannedLeaveAt']),
      approvedAt: j['approvedAt'] == null ? null : parse(j['approvedAt']),
      hasVehicle: j['hasVehicle'] == true,
      plateNo: j['plateNo'] as String?,
      rejectReason: j['rejectReason'] as String?,
      qrToken: j['qrToken'] as String?,
      passcode: j['passcode'] as String?,
    );
  }
}

class VisitorApprovalStep {
  const VisitorApprovalStep({
    required this.action,
    required this.actorType,
    required this.actorName,
    required this.actedAt,
    this.comment,
  });

  final String action;
  final String actorType;
  final String actorName;
  final DateTime actedAt;
  final String? comment;

  factory VisitorApprovalStep.fromJson(Map<String, dynamic> j) => VisitorApprovalStep(
        action: (j['action'] ?? '').toString(),
        actorType: (j['actorType'] ?? '').toString(),
        actorName: (j['actorName'] ?? '').toString(),
        actedAt: j['actedAt'] is String && (j['actedAt'] as String).isNotEmpty
            ? DateTime.tryParse(j['actedAt'] as String) ?? DateTime.now()
            : DateTime.now(),
        comment: j['comment'] as String?,
      );
}
